### PR TITLE
Fixed issue #143 -> Add pre-commit

### PR DESCRIPTION
**Pull request by Anselmoo**
Fixed issue #143

1. Add pre-commit
2. Apply pre-commit to all files
3. Add pre-commit to docs

### CODE REVIEW

**Comment by mpmdean** [comment]
Thanks @Anselmoo. 

To my untrained eye, I would anticipate this being good practice and not something that will cause any issues. Something that jumps out in the file review is that quite a few Fortran and Fortan configuration files get touched in this change. 

Having tried to run some of the Cowan and related codes, I know that Fortran can sometimes be annoyingly sensitive to things like whitespace and new lines. Can @shenmidelin and @mrakitin have a look here? Thanks!

**Comment by Anselmoo** [comment]
> Thanks @Anselmoo.
> 
> To my untrained eye, I would anticipate this being good practice and not something that will cause any issues. Something that jumps out in the file review is that quite a few Fortran and Fortan configuration files get touched in this change.
> 
> Having tried to run some of the Cowan and related codes, I know that Fortran can sometimes be annoyingly sensitive to things like whitespace and new lines. Can @shenmidelin and @mrakitin have a look here? Thanks!

Thx for the quick feedback @mpmdean 

Indeed, we can leave FORTRAN code out! Due to the comment in PR #143, I thought it is a good point to track this in a general way. Especially the python code contains kind of a lot of whitespaces.

What I personally like a lot is the fact, that you can include all linters (black, mypy, pydocstyle, ...) straight into the pre-commit. In this regard, this hook for FORTRAN might be interesting? [https://github.com/pseewald/fprettify](https://github.com/pseewald/fprettify)

**Comment by mpmdean** [comment]
I'd propose to apply the pre-commit to only .py and .rst. I presume that's easily done?  What do other think?

BTW. I believe all our important python passes these. Just ours docs are exempt for now, which is probably lazyness on my part.

**Comment by Anselmoo** [comment]
That's a good idea to start with 
`python`, `rst`, and maybe `JSON `first and include other formats, later. I will exclude the rest out of the general `pre-commit` hook.

File: github_deploy_key_nsls_ii_edrixs.enc [hunk 1]
```diff
@@ -1 +1 @@
...
...n_JkHaBPVCVsaBqu0F-Tz86PGd1dGHcKQJp0RWZRvfQtAS42U0jnTzdbOzDwKCXzykZzXBPW0RUWf9X5yDaROaifOW8TV2RO1-v28gQR8t3qypM2sZsq0YRQV2U5mGVPaiAIUoA-ZrLGMILd3KVVfrISKqlJ3zgfIlG6J6mZIrJjCS4Bb8bDnv22MSGF-FZ7I9n9teuQCBWyiP9USoG3oOMTOxvC5uE6pNAV094nsN1AFhHkVelWOD2WV51nFNcyNtbXFv7yPC502rD7UDmqWAVcZgLEGQLzwkaeWumHxfp_Uig6Ap_DqsYFkHYKx-7f77YBsoqovvPeXQjmHIU_lGzRcJyO8rc2-8BXkTOzfI_ljgEWyw9Gf29wIli9SSFjx_mt983l1AUkmrDSlR9ZgtMEjFQ7sLIvAUpgbXKMA2XwnJsopiXvBXDvwr74q2OEuWHEqNSWTdT25WIOsfo7wEH5iMlOrCeYkt2tocEUwMvyLQoZulIJWNdnt20JacJcEIAfqzN1bZYdShb7d1kLPaKjP6FXf3rwr4KIb6yPM20rfrhKl82UvF7iiBFoeaTnqclhLzmwqTMBg4QN185Omn6S6s7EubEP6YKspsnVXHe7-_IvqjL7SNRVhkcv4E53rL_2MsAbnM1tl9o443SDR2tPVjaqGJpkW2vf2oaFT0Sxqprs2Z2lJHZtq1Tgo9ThBK1xYSLcERZeL2OK2nfskYcuJ_KSjRjYoKKZR_uD2SVD_C1javdfe64wvHHFHm_lpVOaMGSdI5TEwrE_AtHpUGRzKfXsDSHzT2dyfClgL4XdlhC_veRgXecmZQVU9vhtY6rJX-29ldM9P1vt0PMB7nEmW-sH_huTHTPPB-syhVhA4F0IGSKY1UI1AY-aYyvFbwyYaDg_zHDiMbocyX4xQRe2SsqvuujG8YWVJvbZIP3Uhq-ahagCtTdQbMPJtjVWOpQSVKAvQSCMXSihrP97anwqC6cvMLKWnd4ETMy8ZwueK5-z_MNqhYkrhtTSTNLj_yMiJUjnHi7ASE21AZYYqqAeo8oyStH3_FxXaAqVXsNAjFun_DN56wJZNIG9r5YYspM7rnZpdDSfybDdVwmHDnnWYyYXphc0UzRAQIhpzE4RD2V4OgpmQaYdYK207nz_bTfPIjIsjE16UtmNbYfj-23vrKs-0maHuzNE6zxf8KhLeg==
```
Comment (mrakitin) [inline github_deploy_key_nsls_ii_edrixs.enc:1]: This file may not be needed anymore, as the docs are now published by means other than `doctr` (implemented via #125).

**Review by mrakitin — COMMENTED** [review]
Looks good to me. I am happy with the hook applied to the .py and .rst files, not sure about the .f90, as @mpmdean mentioned that can break the code.

@Anselmoo, can you please explain how it's supposed to work? Will the workflow automatically _commit_ the linted changes, or will it just fail and a human will need to fix things manually? I am not currently using any pre-commit hooks, would be good to learn about them more. Thanks!

**Comment by Anselmoo** [comment]
> Looks good to me. I am happy with the hook applied to the .py and .rst files, not sure about the .f90, as @mpmdean mentioned that can break the code.
> 
> @Anselmoo, can you please explain how it's supposed to work? Will the workflow automatically _commit_ the linted changes, or will it just fail and a human will need to fix things manually? I am not currently using any pre-commit hooks, would be good to learn about them more. Thanks!

In general [https://github.com/pre-commit/pre-commit](https://github.com/pre-commit/pre-commit) is python package, which have first install via

1.  `pre-commit install`
2. `pre-commit run --all-files` is good for the first run to check current status quo
3. Next every git-push will run over `pre-commit` and block if the commit does not fulfill the requirements. In the case of `whitespace`, `isort` or `black`, it will automatically reformat. In case of flake8 it will block so long you are not fixing the fail.

If you want, we can add an extended comment in the **README**. I will restore the `*.f90` files and include an exclude.

**Comment by mrakitin** [comment]
Thanks! I think it'd make sense to include the comment about setting up the recommit hooks in the dev section of the docs and/or README.

File: github_deploy_key_nsls_ii_edrixs.enc (hunk 1 above)
Comment (mpmdean) [reply github_deploy_key_nsls_ii_edrixs.enc:1]: I would suggest not changing this in this PR. I will raise an issue about this

File: src/Makefile [hunk 2]
```diff
@@ -40,4 +40,3 @@ clean:
 	rm -f *.x
```
Comment (mpmdean) [inline src/Makefile:40]: I believe that these files should not be changed? As they are in the `src` folder?

File: src/full_diag.f90 [hunk 3]
```diff
@@ -30,4 +30,3 @@ subroutine full_diag_ham(ndim, ham, eval, evec)
```
Comment (mpmdean) [inline src/full_diag.f90:30]: I believe that these files should not be changed? As they are in the `src` folder?

File: src/io.f90 [hunk 4]
```diff
@@ -657,5 +657,3 @@ subroutine write_opavg(n, eval, opavg)
```
Comment (mpmdean) [inline src/io.f90:657]: I believe that these files should not be changed? As they are in the `src` folder?

**Review by mpmdean — CHANGES_REQUESTED** [review]
Thanks @Anselmoo 

There seem to be a couple of files in `src` that still get changed in this PR. Can you check this?

**Comment by Anselmoo** [comment]
Sorry @mpmdean for Late reply. 
As mentioned in the code review, it is end of line trimming. Sometimes it was previously included, sometimes not. Now, it's consistent 
However, if you prefer, I will keeps `src` completely untouched?

**Comment by mpmdean** [comment]
> Sorry @mpmdean for Late reply. As mentioned in the code review, it is end of line trimming. Sometimes it was previously included, sometimes not. Now, it's consistent However, if you prefer, I will keeps `src` completely untouched?

If you don't mind let's leave 
`src/*` 
`docs/user_guide.lyx`
`github_deploy_key_nsls_ii_edrixs.enc`

totally untouched. I'm not knowledgeable enough to know this won't cause problems and since these files are barely read there's not much advantage to cleaning them up.

**Review by mpmdean — APPROVED** [review]

**Comment by mpmdean** [comment]
@shenmidelin can you check and sign off on these changes?

**Pull request merged**